exp/runtime/api/v1beta1/extensionconfig_types.go

```diff
@@ -0,0 +1,195 @@
+	// allowed. Fragments ("#...") and query parameters ("?...") are not
+	// allowed either.
+	//
+	// +optional
+	URL *string `json:"url,omitempty"`
+
+	// Service is a reference to the Kubernetes service for the ExtensionHandler.
+	// Either `service` or `url` must be specified.
+	//
+	// If the ExtensionHandler is running within a cluster, then you should use `service`.
+	//
+	// +optional
+	Service *ServiceReference `json:"service,omitempty"`
+
+	// CABundle is a PEM encoded CA bundle which will be used to validate the ExtensionHandler's server certificate.
+	// +optional
+	CABundle []byte `json:"caBundle,omitempty"`
+}
+
+// ServiceReference holds a reference to a Kubernetes Service.
+type ServiceReference struct {
+	// Namespace is the namespace of the service.
+	Namespace string `json:"namespace"`
+
+	// Name is the name of the service.
+	Name string `json:"name"`
+
+	// Path is an optional URL path which will be sent in any request to
+	// this service. If a path is set it will be used as prefix and the hook-specific
+	// path will be appended.
+	// +optional
+	Path *string `json:"path,omitempty"`
+
+	// Port is the port on the service that hosting ExtensionHandler.
+	// Default to 8443.
+	// `port` should be a valid port number (1-65535, inclusive).
+	// +optional
+	Port *int32 `json:"port,omitempty"`
+}
+
+// ANCHOR_END: ExtensionConfigSpec
+
+// ANCHOR: ExtensionConfigStatus
+
+// ExtensionConfigStatus defines the observed state of ExtensionConfig.
+type ExtensionConfigStatus struct {
+	// ExtensionHandlers defines the current ExtensionHandlers supported by an Extension.
+	// +optional
+	ExtensionHandlers []ExtensionHandler `json:"extensionHandlers,omitempty"`
+
+	// Conditions define the current service state of the ExtensionConfig.
+	// +optional
+	Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+}
+
+// ExtensionHandler specifies the details of a handler for a particular runtime hook registered by an Extension server.
+type ExtensionHandler struct {
+	// Name is the unique name of the ExtensionHandler.
+	Name string `json:"name"`
+
+	// Hook defines the versioned runtime hook which this ExtensionHandler serves.
+	Hook GroupVersionHook `json:"hook"`
+
+	// TimeoutSeconds defines the timeout duration for client calls to the ExtensionHandler.
+	// +optional
+	TimeoutSeconds *int32 `json:"timeoutSeconds,omitempty"`
+
+	// FailurePolicy defines how failures in calls to the ExtensionHandler should be handled by a client.
+	// +optional
+	FailurePolicy *FailurePolicy `json:"failurePolicy,omitempty"`
+}
+
+// GroupVersionHook defines the runtime hook when the ExtensionHandler is called.
+type GroupVersionHook struct {
+	// APIVersion is the Version of the Hook.
+	APIVersion string `json:"apiVersion"`
+
+	// Hook is the name of the hook.
+	Hook string `json:"hook"`
+}
+
+// FailurePolicy specifies a failure policy that defines how unrecognized errors from the admission endpoint are handled.
+type FailurePolicy string
```

Comment:
This field represents an enum, you should add an enum kubebuilder tag, then when it's included in an API, you'll get the validation
```suggestion
// +kubebuilder:validation:Enum:=Ignore;Fail
type FailurePolicy string
```